Entity rideable feature

Studgi opened the pull request:
This PR is an updated version of #330 but I made several changes according to the feedback given in that PR.

- If a rideable has 2 or more riders; if any dismounts occurred, the API should not be responsibly for reordering the seat index of the remaining rider- it should be changed outside of dragonfly.
- It's much simpler and voids confusion between <Rideable>.AddRider() and <Rider>.MountEntity()
- It's still very flexible since you can change the seat position during runtime via the HandleMount handler (seatPos *Vec3)
- You are able to control who the driver is without using their seat position to do so

If you want to suggest changes, please let me know.

Studgi committed:
Merge pull request #1 from df-mc/master

session: Set `EmoteFlagMuteChat` if emote chat muted. (#1116)

Studgi committed:
Refactor entity riding system and update dependencies

- Downgrade Go version from 1.24 to 1.25.0 to match project requirements
- Update all dependency versions to latest compatible releases
- Completely redesign entity riding interface with proper Rider and Rideable contracts
- Implement mount/dismount handlers in player handler interface
- Add rider state tracking in player data structure
- Implement entity metadata synchronization for riding state
- Handle vehicle interactions in inventory transaction system
- Propagate movement inputs from riders to ridden entities
- Add viewer methods for entity mount/dismount events
- Ensure proper cleanup on player close with automatic dismount

Studgi committed:
Standardize vector types in riding system

- Change seat position vectors from mgl64.Vec3 to mgl32.Vec3 for protocol compatibility
- Update all method signatures and field types to use mgl32.Vec3 consistently

Studgi committed:
Downgrade to go 1.24.7

Studgi committed:
revert go version and modules bump

TwistedAsylumMC reviewed (changes requested):
Thank you for picking up on the old PR, it's very appreciated. I've left a few comments in the code but I do have one other concern.

There's currently no way for the ridden entity to control how many riders it can have but also where the seats are, which I feel like it's something that should happen. From what I can tell, any player can go up to the same entity and start riding it, ending up in the same place as everyone else unless a player handler changes this? Rideable() should most likely have some sort of method that allows controlling this, providing the seat positions etc.

Studgi commented on server/entity/riding.go:
```diff
@@ -0,0 +1,26 @@
+	// RidingEntity returns the entity the player is currently riding.
+	RidingEntity() Rideable
+	// SeatPosition returns the position of where the player is sitting.
+	SeatPosition() mgl32.Vec3
```
Ohh okay

Studgi committed:
Refactor entity riding system with seat-based approach

- Replace seat position vectors with seat index system for better protocol compatibility
- Implement seat management with NextFreeSeatIndex method on Rideable entities
- Add seat position lookup method that calculates position from seat index
- Change mount/dismount handler signatures to use seat index instead of position
- Update entity metadata to use calculated seat positions instead of stored vectors
- Maintain driver status based on seat index 0 (first seat controls the entity)
- Simplify riding logic by removing direct position manipulation
- Ensure proper viewer updates when changing seats or dismounting

Studgi committed:
Merge branch 'master' into master

DaPigGuy commented on server/session/handler_inventory_transaction.go:
```diff
@@ -158,6 +159,11 @@ func (h *InventoryTransactionHandler) handleUseItemOnEntityTransaction(data *pro
 	switch data.ActionType {
 	case protocol.UseItemOnEntityActionInteract:
+		if rideable, ok := e.(entity.Rideable); ok {
+			if nextSeatIndex, ok := rideable.NextFreeSeatIndex(); ok {
```
Could we pass the ClickedPosition so that it is possible to allow choosing a seat? e.g. the train in the Hive's hub

Studgi replied:
Ofc bbg will do today

Studgi committed:
Add click position parameter to seat selection

- Modify NextFreeSeatIndex to accept click position for context-aware seat selection
- Pass clicked position from interaction data to seat selection method

Studgi committed:
Merge remote-tracking branch 'origin/master'

TwistedAsylumMC commented on server/player/player.go:
```diff
@@ -54,6 +54,9 @@ type playerData struct {
 	armour                       *inventory.Armour
 	heldSlot                     *uint32
 
+	rideable  entity.Rideable
```
I'd call this "riding" instead of "rideable", it currently doesn't sound like it's the entity the player is riding

TwistedAsylumMC replied:
Still not clear imo, I think you need to have "riding" in there somewhere. If you want you can also add a doc to explain but i don't think that's necessary with just "riding" or "ridingEntity". You could also do "riddenEntity" or something like that. The "able" just sounds like it's something that can be ridden, not actively being ridden

TwistedAsylumMC reviewed (changes requested):
I'm happy with the direction this has gone, just a few minor comments. I assume this has been tested with the most recent changes? Would be nice to see some minimal code showing an example entity being created with seats etc.

Studgi commented:
> I'm happy with the direction this has gone, just a few minor comments. I assume this has been tested with the most recent changes? Would be nice to see some minimal code showing an example entity being created with seats etc.
Yep it's tested. Also where do I put the example code?

TwistedAsylumMC commented on server/player/player.go:
```diff
@@ -2554,6 +2557,78 @@ func (p *Player) SetMaxAirSupply(duration time.Duration) {
+	if p.rideable == nil || p.seatIndex == -1 || len(p.rideable.SeatPositions()) <= p.seatIndex {
+		return mgl64.Vec3{}, false
+	}
+	return p.rideable.SeatPositions()[p.seatIndex], true
```
Yes you're completely right, I am just blind

Studgi committed:
renamed rideable to riddenEntity

Studgi committed:
Merge branch 'master' into master

Studgi committed:
Merge pull request #2 from df-mc/master

update from master

Studgi committed:
riding: add methods for managing multiple riders on riddenEntity interface

Studgi committed:
Merge remote-tracking branch 'origin/master'

Studgi committed:
riding: add comprehensive developer documentation for Rider and Rideable interfaces

Studgi committed:
sleep: implement sleep functionality and related mechanics

Studgi committed:
entity_metadata: add support for baby entities in metadata flags

Studgi committed:
entity_metadata: fix flag setting for baby entities in metadata

Studgi committed:
entity_metadata: rename IsBaby method to Baby for consistency

Studgi commented on server/session/entity_metadata.go:
```diff
@@ -70,7 +70,7 @@ func (s *Session) addSpecificMetadata(e any, m protocol.EntityMetadata) {
 		m.SetFlag(protocol.EntityDataKeyFlags, protocol.EntityDataFlagInvisible)
 	}
 	if b, ok := e.(baby); ok && b.IsBaby() {
```
Several other people explained why this was needed for consistency, so I updated the code. Thanks for pointing this out!

Studgi committed:
player: refactor riding entity handling to use world transactions

Studgi committed:
refactor interaction handlers to utilize world transactions

Studgi committed:
player: update RidingEntity method to return a boolean for entity existence

Studgi committed:
player: update RidingEntity method to return a boolean for existence check

Studgi committed:
player: modify RidingEntity method to return nil for non-existent entities

Studgi committed:
Merge pull request #3 from df-mc/master

update

Studgi committed:
remove unused baby interface from entity metadata

Studgi committed:
Merge pull request #4 from df-mc/master

Fix bugs, improve performance, and enhance block implementations

Studgi committed:
Merge pull request #5 from df-mc/master

Update dragonfly for version 1.26.10 and add Within() method

Studgi committed:
merge with upstream

HashimTheArab commented:
updated this to latest df and fixed some issues
